src/app/Shared/Services/Login.service.tsx

```diff
@@ -188,23 +187,66 @@ export class LoginService {
     return this.logout.asObservable();
   }
 
-  setLoggedOut(): void {
-    this.removeCacheItem(this.USER_KEY);
-    this.removeCacheItem(this.TOKEN_KEY);
-    this.token.next('');
-    this.username.next('');
-    this.logout.next();
-    this.sessionState.next(SessionState.NO_USER_SESSION);
-
-    // TODO implement a logout call that will delete the access_token
-    // cached in the oauth server, thus redirecting users to the login page
-    this.queryAuthMethod();
+  setLoggedOut(token: string, method: string): Observable<boolean> {
+    return fromFetch(`${this.authority}/api/v2.1/logout`, {
+      credentials: 'include',
+      mode: 'cors',
+      method: 'POST',
+      body: null,
+      headers: this.getAuthHeaders(token, method),
+    })
+    .pipe(
+      switchMap(response => {
+        if(response.status === 302) {
+          const redirectUrl = response.headers.get('X-Location') || '';
+          return this.logoutOAuthServer(redirectUrl, token, method);
+        } else {
+          this.resetSessionState();
+          return of(true);
+        }
+    }),
+    catchError((e: Error): ObservableInput<boolean> => {
+      window.console.error(JSON.stringify(e));
+      return of(false);
+      }),
+    );
   }
 
   setSessionState(state: SessionState): void {
     this.sessionState.next(state);
   }
 
+  private logoutOAuthServer(redirectUrl: string, token: string, method: string): Observable<boolean> {
+    return fromFetch(redirectUrl, {
+        credentials: 'include',
+        mode: 'no-cors',
+        method: 'POST',
+        body: null,
+        headers: this.getAuthHeaders(token, method),
+      }).pipe(
+          map(response => {
+            this.resetSessionState();
+            this.navigateToLoginPage();
+            return true;
```

Comment:
Should this be mapping `response => true` or perhaps `response => response.ok`?

The `resetSessionState()` and `navigaeToLoginPage()` calls would fit better stylistically in a `tap` than a `map`. Perhaps the `tap` should be after the map to boolean or even after the `catchError`, and these actions only performed if the request succeeded?

Reply:
Since I realized `mode: no-cors` prevents the frontend from accessing the response data, I found some docs explaining how to use CORS with the OpenShift server. `config.openshift.io` lets you edit the spec `additionalCORSAllowedOrigins` ([APIServer docs](https://docs.openshift.com/container-platform/4.8/rest_api/config_apis/apiserver-config-openshift-io-v1.html)). 

When I tried editing the APIServer config manually ([OCP docs](https://access.redhat.com/documentation/en-us/openshift_container_platform/4.4/html/security/allowing-javascript-based-access-api-server)), I can access the 200 `response.ok` status, but the `kube:admin` session cookie is still stored on the OAuth server. Pressing logout right now immediately logs you back in with a new token, regardless of whether I send `headers: this.getAuthHeaders(...)` with my request.

```
$ oc edit apiserver.config.openshift.io cluster

apiVersion: config.openshift.io/v1
kind: APIServer
metadata:
  annotations:
    include.release.openshift.io/ibm-cloud-managed: "true"
    include.release.openshift.io/self-managed-high-availability: "true"
    include.release.openshift.io/single-node-developer: "true"
    oauth-apiserver.openshift.io/secure-token-storage: "true"
    release.openshift.io/create-only: "true"
  creationTimestamp: "2021-12-17T21:24:58Z"
  generation: 4
  name: cluster
  resourceVersion: "1147467"
  uid: a6c97a3c-d651-48a8-80b0-3d10cfb58ef3
spec:
  additionalCORSAllowedOrigins:
  - cryostat-sample-default\.apps\.jalaw0\.lab\.upshift\.rdu2\.redhat\.com
  audit:
    profile: Default
```